core/Config.php

```diff
@@ -514,7 +514,7 @@ protected function getUncached($class, $name, $sourceOptions, &$result, $suppres
 		// Include extensions only if not flagged not to, and some have been set
 		if (($sourceOptions & self::EXCLUDE_EXTRA_SOURCES) != self::EXCLUDE_EXTRA_SOURCES) {
 			// If we don't have a fresh list of extra sources, get it from the class itself
-			if (!array_key_exists($class, $this->extraConfigSources)) {
+			if (!isset($class[$this->extraConfigSources]) || $class[$this->extraConfigSources] === null) {
```

Comment:
I actually realized my mistake late yesterday with this optimization when I was playing with 'hasField' optimizations. Planning to use instead an isset() followed by array_key_exists(). 

ie. !isset($class[$this->extraConfigSources]) || !array_key_exists($class, $this->extraConfigSources)

As that should give a speed boost in all not variable !== null cases. I'm probably going to experiment with writing an optimizer in PHP that just converts 'is_array' into the cast optimized version as well as this, as that should improve the floor speed of Silverstripe by a few noticeable ms.